### ISSUE TITLE
Rework file handling for users/privileges

### ISSUE THREAD

**Issue by Hydrocharged** [dolthub/dolt#3711]
Currently, using users and privileges requires an explicit declaration of the `mysql.db` file via `--privilege-file`. This should be changed so that we generate a hidden `.dolt.mysql.db` file. The steps for this process are as follows (when running `dolt sql` or `dolt sql-server` from a directory):

1. If the `--mysql-db-file` argument is provided, then use the file found there.
    * Create a file if one was not found.
2. If the `--privilege-file` argument is provided, display a warning that the argument is deprecated, then use the file found there.
    * Create a file if one was not found.
3. Check all subdirectories of the current directory for a `.dolt.mysql.db` file.
    * If a file is found, throw an error.
    * No need to recursively check all subdirectories.
4. Check the parent directory for a `.dolt.mysql.db` file and the current directory for a `.dolt.mysql.db` file.
    * If only one of the files exists, then use that file.
    * If both files exist, then throw an error.
5. Check the current directory for a `mysql.db` file.
    * If the file exists, only read from it and create a hidden `.dolt.mysql.db` file in the current directory as it will be written to.
6. Check the current directory for a `privs.json` file.
    * If the file exists, only read from it and create a hidden `.dolt.mysql.db` file in the current directory as it will be written to.
7. Check the parent directory for a `mysql.db` file.
    * If the file exists, only read from it and create a hidden `.dolt.mysql.db` file in the parent directory as it will be written to.
8. Check the parent directory for a `privs.json` file.
    * If the file exists, only read from it and create a hidden `.dolt.mysql.db` file in the parent directory as it will be written to.
9. Create a hidden `.dolt.mysql.db` file in the current directory and use it.

At each step, if we have a file to use or we throw an error, we do not proceed to the steps below.

When reading a file, we should first attempt to read it as a FlatBuffers-encoded file. If that fails, we then try it in the legacy JSON-encoding. This should cover all current cases and allow for compatibility with existing workflows. We should also hide the `--privilege-file` argument from the CLI help and online documentation, as the argument name is no longer correct for what the file actually represents. `--mysql-db-file` is the better argument name, and the one we should display going forward.

After the above changes are completed, we need to update our online documentation, as it is completely out-of-date.

https://docs.dolthub.com/sql-reference/server/access-management
https://docs.dolthub.com/cli-reference/cli#dolt-sql-server

**Comment by zachmu** [comment]
Daylon and I synced up on this offline, here's a summary of what we concluded:

* I don't like the single `mysql.db` file, Daylon doesn't like the idea of multiple files littering a directory when the server is run with default settings. Compromise is to store all persisted data of this nature in as many files as we think appropriate, in a `.doltcfg` directory. Inside this directory will live the `privileges.db` file and any other files of a similar nature we decide to store, and serve as a general purpose place to put persistent data for a server.
* `.doltcfg` gets created on demand if it doesn't exist, and its default location is `$datadir/.doltcfg`. We introduce a new flag, `--doltcfg-dir`, that allows you to override this location.
* `--privileges-file` can be specified as well, and its default is `$doltcfg/privileges.db`

There's a single error / sanity check for users who have run dolt both from within a dolt database directory and from its parent directory: if we detect that dolt was invoked in a dolt database directory, and none of `--data-dir`, `--doltcfg-dir`, `--privileges-file`  were specified, then error if we find a `.doltcfg` directory in both the working directory and its parent.

Otherwise, if none of these flags are given, look for `.doltcfg` in the parent directory before creating it in the current directory.

This sounds complicated, but there's only ambiguity in the case that no config flags are passed to `dolt sql` or `dolt sql-server`.

To summarize, here are all flags and their defaults:

* `data-dir`: `.`
* `doltcfg-dir`: `$data-dir/.doltcfg`
* `privilege-file`: `$doltcfg-dir/privileges.db`

**Comment by JCOR11599** [comment]
In the case where there is a `.doltcfg` in both the working and parent directory, does the user fix it by deleting one of the `.doltcfg` directories?

**Comment by zachmu** [comment]
In cases of ambiguity, the error message should indicate that you can resolve it by setting `--doltcfg-dir` to one of the two locations, and list them.
